src/connector.rs

```diff
@@ -0,0 +1,563 @@
+use crate::fungible_token::*;
+use crate::parameters::*;
+use crate::sdk;
+use crate::types::*;
+
+use crate::deposit_event::*;
+use crate::engine::Engine;
+use crate::prelude::*;
+use crate::prover::validate_eth_address;
+#[cfg(feature = "log")]
+use alloc::format;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+pub const CONTRACT_NAME_KEY: &str = "EthConnector";
+pub const CONTRACT_FT_KEY: &str = "ft:";
+pub const NO_DEPOSIT: Balance = 0;
+const GAS_FOR_FINISH_DEPOSIT: Gas = 50_000_000_000_000;
+const GAS_FOR_VERIFY_LOG_ENTRY: Gas = 40_000_000_000_000;
+const GAS_FOR_TRANSFER_CALL: Gas = 40_000_000_000_000;
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct EthConnectorContract {
+    contract: EthConnector,
+    ft: FungibleToken,
+}
+
+/// eth-connector specific data
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct EthConnector {
+    pub prover_account: AccountId,
+    pub eth_custodian_address: EthAddress,
+}
+
+/// Token message data
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub enum TokenMessageData {
+    Near(AccountId),
+    Eth { address: AccountId, message: String },
+}
+
+/// On-transfer message
+pub struct OnTransferMessageData {
+    pub relayer: AccountId,
+    pub recipient: EthAddress,
+    pub fee: U256,
+}
+
+impl EthConnectorContract {
+    pub fn get_instance() -> Self {
+        Self {
+            contract: sdk::get_contract_data(CONTRACT_NAME_KEY),
```

Comment:
Much like https://github.com/aurora-is-near/aurora-engine/pull/62#discussion_r627449415 I suggest we use the `KeyPrefix` enum rather than ad-hoc strings to create unique storage keys.

For example we could add `KeyPrefix::ContractData` and then change the definition of `sdk::get_contract_data` to be something like

```rust
enum ContractId {
    EthConnector = 0x0,
    FungibleToken = 0x1,
}

fn get_contract_data<T: BorshDeserialize>(contract: ContractId) -> T {
    let key = [KeyPrefix::ContractData as u8, contract as u8];
    let data = read_storage(&key).expect("Failed read storage");
    T::try_from_slice(&data).unwrap()
}
```

Reply:
@birchmd changed to your proposal

Reply:
Thanks for getting that change in so quickly!